### DISCUSSION TITLE
Learning TipTap - How to make Custom Extentions

### DISCUSSION THREAD

**Discussion by CarelessCourage** [ueberdosis/tiptap#769] [General Questions]
Tiptap itself is all easy enough to understand. But I find that the documentation for how to build your own extensions is not just lacking but literally nonexistent. I figured I could learn by reading the prosemirror docs, but all I seem to be able to get from that is some familiarity with the underlying concepts and terms. It doesn't really tell me much in terms of how to build a TipTap extension.

I have been able to get a lot of stuff working myself by just looking through random questions others have asked and looking through examples and such and just experimenting by trying things our. This codesandbox was particularly helpful in teaching me how to work with it: [sandbox](https://codesandbox.io/s/9x8fr?file=/src/components/CustomStyle.js)

I learned enough from this that I can now make some basic extension myself. But i still need to have a better understanding of how it all works if I want to have greater freedom in using tiptap. 

like, for instance. How does the get schema function work in tiptap? And how do I call the getAttrs function in this code? And what exactly does parseDOM do here? And whats the importance of attrs? Also, how does the commands function work? 

![Screenshot (51)](https://user-images.githubusercontent.com/12764398/88241903-d5efbb80-cc8b-11ea-9379-869ad13d58d0.png)

As a newbie, I have a lot of questions, and trying to decipher how it works from bits and pieces scattered around in other peoples code, although its slowly, and painfully, getting me where I want to be, it is very slow, and frequently frustrating way of learning. 

My question is, is there any tutorials or documentations, or resources out there that I should be looking at to get a good start? Or have you all just figured it out with guesswork and prior coding knowledge from other, similar things? Is there really no documentation about any of this anywhere on the internet? Is this all just super obvious and I'm just lacking some more fundamental coding understanding? In which case, which understanding should I be brushing up on?

**Comment by hanspagel** [comment]
Thanks for your post! 

No worries, it’s *not* super obvious and you’re right the documentation is not very helpful. We didn’t expect to have so many people using tiptap and are very thankful for the community that evolved around tiptap. Great to hear that you could learn from questions, sandboxes and the ProseMirror documentation, but you’re right. There is a lot potential to make this easier.

That said, we started to actively maintaining the repository again. That means, we’re working through questions, issues and pull requests. And we plan to work on a new version. To be honest, there is already a proof of concept and that includes a lot new things. One of them is an interactive documentation. With examples, code and explanations. What I try to say: We’re working hard to improve that.

If you want to help us, there are a few ways:
1. Feel free to share your questions with us, so we know what people want to know.
2. Consider sharing solutions you’ve found with us. Do you have experience making a PR? If you like, you can send a PR to the `ueberdosis/tiptap-docs` repository with your solutions. No worries, if you think it’s not perfect. I’ll try my best to describe it better or whatever.
3. We just set up the GitHub sponsorship program for our organization and got already 11 people sponsoring us. That’s a huge motivation for us, to invest more time in tiptap. Consider to sponsor us too. https://github.com/sponsors/ueberdosis

Thanks again for sharing, and like I said: No worries. It’s not your fault. I’m impressed on how far you’ve come with your research. Keep it up!

**Comment by isaac-souza** [comment]
> Tiptap itself is all easy enough to understand. But I find that the documentation for how to build your own extensions is not just lacking but literally nonexistent. I figured I could learn by reading the prosemirror docs, but all I seem to be able to get from that is some familiarity with the underlying concepts and terms. It doesn't really tell me much in terms of how to build a TipTap extension.
> 
> I have been able to get a lot of stuff working myself by just looking through random questions others have asked and looking through examples and such and just experimenting by trying things our. This codesandbox was particularly helpful in teaching me how to work with it: [sandbox](https://codesandbox.io/s/9x8fr?file=/src/components/CustomStyle.js)
> 
> I learned enough from this that I can now make some basic extension myself. But i still need to have a better understanding of how it all works if I want to have greater freedom in using tiptap.
> 
> like, for instance. How does the get schema function work in tiptap? And how do I call the getAttrs function in this code? And what exactly does parseDOM do here? And whats the importance of attrs? Also, how does the commands function work?
> 
> ![Screenshot (51)](https://user-images.githubusercontent.com/12764398/88241903-d5efbb80-cc8b-11ea-9379-869ad13d58d0.png)
> 
> As a newbie, I have a lot of questions, and trying to decipher how it works from bits and pieces scattered around in other peoples code, although its slowly, and painfully, getting me where I want to be, it is very slow, and frequently frustrating way of learning.
> 
> My question is, is there any tutorials or documentations, or resources out there that I should be looking at to get a good start? Or have you all just figured it out with guesswork and prior coding knowledge from other, similar things? Is there really no documentation about any of this anywhere on the internet? Is this all just super obvious and I'm just lacking some more fundamental coding understanding? In which case, which understanding should I be brushing up on?

Thanks a lot for sharing this, when I saw your parseDOM method, I immediately knew what was missing in my implementation and it helped me solve something that was annoying me so freaking much
